Migrate some calls to FHIR and GraphQL

bmonjoie opened the pull request:
The new implementation relies on OkHttp and Apollo. 
OkHttp was added because the current implementation relies on Apache HTTP client, which is not supported anymore since Android 6 (API 23). A version supports Android 23+, but Apache does not officially maintain it. Also, OkHttp is the current standard and comes with lots of benefits, like, among other things, the logger, automatic support for GZip, and a strong security focus.
Apollo is in its version 2 as version 3+ requires Kotlin Coroutines. Also, the plan is that some new endpoints or extensions could be added to the FHIR API to remove the dependency on GraphQL.

Everything is hidden behind use cases to impact the current code base as little as possible.

Also, the code was migrated to AndroidX because of Apollo's dependencies. 
The versions of the AndroidX have been kept back to avoid conflicts and upgrades.

dragos-dobre committed:
Finish release/v2.0.0

Finish release/v2.0.0

dragos-dobre committed:
Finish release/v2.0.1

Finish release/v2.0.1

dragos-dobre committed:
Finish Release/v2.0.1

Finish Release/v2.0.1

edarchis committed:
Github Action: manual build

dragos-dobre committed:
Merge pull request #50 from openimis/manual-build-action

Github Action: manual build

dragos-dobre committed:
Finish release/v2.1.0

Release v2.1.0

dragos-dobre committed:
Update activity_about.xml

malinowskikam committed:
Updated java version in github actions

dragos-dobre committed:
Finish feature/github_action_update

Github actions Java update

malinowskikam committed:
github actions: Changes cache from maven to gradle

dragos-dobre committed:
Finish feature/github_actions_fix

Github actions fix

dragos-dobre committed:
Merge pull request #80 from openimis/develop

Develop

dragos-dobre committed:
Finish release/2.2.0

Release/2.2.0

dragos-dobre committed:
Finish feature/OTC-745

OTC-745 Implemented claim 2-way sync

malinowskikam committed:
Revert "OTC-745 Implemented claim 2-way sync"

delcroip committed:
Merge pull request #85 from openimis/revert-83-feature/OTC-745

Revert "OTC-745 Implemented claim 2-way sync"

bmonjoie committed:
Migrate some calls to FHIR and GraphQL (WIP)

The new implementation relies on OkHttp and Apollo.
OkHttp was added because the current implementation relies on Apache HTTP
client, which is not supported anymore since Android 6 (API 23).
A version supports Android 23+, but Apache does not officially maintain it.
Also, OkHttp is the current standard and comes with lots of benefits like,
among other things, the logger, automatic support for GZip, and a strong
security focus.
Apollo is in its version 2 as version 3+ requires Kotlin Coroutines.
Also, the plan is that some new endpoints or extensions could be added to the
FHIR API to remove the dependency on GraphQL.

Everything is hidden behind use cases to impact as less as possible the
current code base.

Also, migrate to AndroidX because of Apollo's dependencies. The versions
of the AndroidX have been kept back to avoid conflicts and upgrades.

bmonjoie commented:
This PR depends on: 
- https://github.com/openimis/openimis-be-api_fhir_r4_py/pull/131 to be merged.
- https://github.com/toch/openimis-be-api_fhir_r4_py/tree/add_hf_code_to_practitioner to be merged

bmonjoie committed:
Moving the claim creation to FHIR (WIP)

This depends on a modification of the backend which would allow
to pass a reference with an OpenIMIS Code rather than a UUID.

There is still an open question about the "REJECTED" status to be solved.

OMT-339

bmonjoie committed:
Moving GetControls to GraphQL

Deleting `ToRestApi` and removing Apache HTTP dependencies.
Adding a way to change the "Login/Logout" button automatically on login.

This depends on a new module migrating the controls to GraphQL that has not
yet been merged.

OMT-339

bmonjoie committed:
Improve "Add claim" UI

dragos-dobre commented:
The second link you mention is a branch and not a PR.

toch commented:
Benjamin told me that the main changes are in directories `domain`, `network`,  and  `usecase`. He advices to start with `network`. As I'm working on the backend to support some of the needs, I'll go through this PR too.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/dto/ActivityDefinitionDto.java:
```diff
@@ -0,0 +1,127 @@
+
+    @NonNull
+    public static List<ActivityDefinitionDto> fromJson(@NonNull JSONObject jsonObject) throws JSONException, ParseException {
+        ArrayList<ActivityDefinitionDto> list = new ArrayList<>();
+        JSONArray entries = jsonObject.getJSONArray("entry");
+        for (int i = 0; i < entries.length(); i++) {
+            JSONObject entry = entries.getJSONObject(i).getJSONObject("resource");
+            JSONArray extensions = entry.getJSONArray("extension");
+            JSONObject value = null;
+            for (int j = 0; j < extensions.length(); j++) {
+                JSONObject extension = extensions.getJSONObject(j);
+                if (extension.has("valueMoney")) {
+                    value = extension.getJSONObject("valueMoney");
+                    break;
+                }
+            }
+            Objects.requireNonNull(value);
+            list.add(
+                    new ActivityDefinitionDto(
+                            /* id = */ entry.getString("id"),
+                            /* identifiers = */ IdentifierDto.fromJson(entry.getJSONArray("identifier")),
+                            /* name = */ entry.getString("name"),
+                            /* title = */ entry.getString("title"),
+                            /* price = */ value.getDouble("value"),
+                            /* currency = */ value.getString("currency"),
+                            /* status = */ entry.getString("status"),
+                            /* date = */ Objects.requireNonNull(DATE_FORMAT.parse(entry.getString("date")))
+                    )
+            );
+        }
+        return list;
```
As a possible future improvement, what do you think of sharing schemas so we can validate the document before parsing it?

For deserialization itself, why have you chosen an explicit approach rather than using a library such as jackson?

bmonjoie replied:
Having JSON schema and/or Swagger documentation would be great. At the moment, I couldn't find any.

The choice of not using a library was made for the following reasons:
- With the current application state, I didn't want to add dependencies that could add "black magic" to the code and make it harder to maintain for someone not versed in the Android ecosystem.
- The FHIR format doesn't always have a clean JSON structure which would require me to write my own parsers anyway
- Going for the explicit approach makes reviewing what is mapped to what easier.

toch replied:
FYI, for FHIR, they provide a light [swagger](https://demo.openimis.org/api/api_fhir_r4/docs/swagger/), [redoc](https://demo.openimis.org/api/api_fhir_r4/docs/redoc/), and the raw [OpenAPI 3 YAML spec](https://demo.openimis.org/api/api_fhir_r4/docs/).

:100: :+1: for being explicit.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/dto/DiagnosisDto.java:
```diff
@@ -0,0 +1,53 @@
+
+    @NonNull
+    public static List<DiagnosisDto> fromJson(@NonNull JSONObject jsonObject) throws JSONException, ParseException {
+        List<DiagnosisDto> diagnoses = new ArrayList<>();
+        JSONArray jsonDiagnoses = jsonObject.getJSONArray("concept");
+        for (int i = 0; i < jsonDiagnoses.length(); i++) {
+            JSONObject diagnosis = jsonDiagnoses.getJSONObject(i);
+            diagnoses.add(
+                    new DiagnosisDto(
+                            diagnosis.getString("code"),
+                            diagnosis.getString("display")
+                    )
+            );
+        }
```
what do you think of a "Stream" way (code not tested):

```java
List<DiagnosisDto> diagnoses = 
    jsonObject
    .getJSONArray("concept")
    .stream()
    .map(
        diagnosis -> 
        new DiagnosisDto(
            diagnosis.getString("code"),
            diagnosis.getSTring("display")))
    .collect(Collectors.toList());
```

bmonjoie replied:
Android 5.0 (the minimum version supported currently) doesn't support streams out of the box. There are ways to backport it to Android. Yet, I didn't want to add too many dependencies for the same reason as exposed above.

bmonjoie replied:
That's also why I didn't introduce Kotlin to the code base

toch replied:
TIL, :bow:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/dto/IdentifierDto.java:
```diff
@@ -0,0 +1,67 @@
+
+    @NonNull
+    public static String getCode(@NonNull List<IdentifierDto> identifiers) {
+        for (IdentifierDto identifier : identifiers) {
+            if (identifier.getType().equals(TYPE_CODE)) {
+                return identifier.getValue();
+            }
+        }
+        throw new IllegalStateException("Couldn't find a value for '"+TYPE_CODE+"'");
```
Another suggestion of a stream way (code not tested):

```java
return identifiers.stream()
  .filter(identifier -> identifier.getType().equals(TYPE_CODE))
  .findFirst()
  .orElseThrow(() -> new IllegalStateException("Couldn't find a value for '"+TYPE_CODE+"'"))
  .getValue();
```

bmonjoie replied:
See answer above

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/dto/LoginDto.java:
```diff
@@ -0,0 +1,26 @@
+    @NonNull
+    private final String password;
+
+    public LoginDto(@NonNull String username, @NonNull String password) {
```
It just appears to me that not every DTO are offering a ` fromJson`  class method . There might be a good reason to do it for `ActivityDefinitionDto`  but not the others. Could you tell me what's your reasons?

From my point of view I can see pros and cons to both approaches. However for the sake of consistency, I would stick to one or the other and not mixing them.

bmonjoie replied:
The `LoginDto` is an object that we are sending to the backend, for which values are not provided to me in JSON.

Creating a JSON to pass it to the object to then parse the values I already have would be counterproductive.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/dto/ActivityDefinitionDto.java:
```diff
@@ -0,0 +1,127 @@
+    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US);
+
+    @NonNull
+    public static List<ActivityDefinitionDto> fromJson(@NonNull JSONObject jsonObject) throws JSONException, ParseException {
```
I find the name a little bit misleading. When used `ActivityDefinitionDto.fromJson(...)`, I would expect at first that it returns an object of `ActivityDefinitionDto`. I'd go for something like `ActivityDefinitionDtoList.fromJson` or `ActivityDefinitionDto.fromJsonToList`.

wdyt?

bmonjoie replied:
Since Java is statically typed and these are clearly utility methods, I don't find it confusing, and making the name longer would not make the code cleaner, IMHO.

toch replied:
characters are not that expensive that you have to save them ;).

anyway fine by me.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/dto/PractitionerDto.java:
```diff
@@ -0,0 +1,174 @@
+            for (int j = 0; j < jsonQualifications.length(); j++) {
+                qualifications.add(CodeDto.fromJson(jsonQualifications.getJSONObject(j).getJSONObject("code")));
+            }
+            String healthFacilityCode = null;
```
Wouldn't you enforce its presence and throw an exception otherwise?

bmonjoie replied:
"Practitioner" does not equal "Claim Admin". It is my understanding that it is completely valid for a Practitioner not to be linked to a health facility.
The check for nullability is made in the use case `FetchClaimAdmins`

toch replied:
you're right, you have enrollment officer.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/request/GetPractitionerRolesRequest.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.List;
+
+public class GetPractitionerRolesRequest extends BaseFHIRGetPaginatedRequest<PractitionerRoleDto> {
```
it seems it's not used, if it's the case, I'd remove it.

bmonjoie replied:
Yes, I made that one when the extension was not available to retrieve the health facility code directly. I haven't removed it but I probably should

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/dto/PractitionerRoleDto.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@SuppressWarnings("unused")
```
if if's unused (and it seems it's the case), I'd remove it.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/okhttp/AuthorizationInterceptor.java:
```diff
@@ -0,0 +1,41 @@
+    @NonNull
+    @Override
+    public Response intercept(@NonNull Chain chain) throws IOException {
+        Token token = global.getJWTToken();
```
(out of the scope of the present PR)

Does it manage the expiration?

bmonjoie replied:
The `Token` class (which I haven't changed except to add a utility method) does take into consideration the expiration date

bmonjoie replied:
The goal of this class is to add the `header` if the token is present and valid. 
As you can see on the next line, there is a `if (token != null && token.isTokenValidJWT()) {`

toch commented on claimManagement/src/main/res/drawable/ic_add.xml:
```diff
@@ -0,0 +1,5 @@
+<vector android:height="24dp" android:tint="#000000"
```
I haven't reviewed those XML files as my knowledge is outdated regarding those Android stuff.

bmonjoie replied:
Also, this was generated by Android Studio based on Google's Material's icon. I don't expect anyone to review those :wink:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/util/Mapper.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @NonNull
+    public List<U> map(@NonNull List<T> list) {
```
I've seen its usage in `claimManagement/src/main/java/org/openimis/imisclaims/network/util/PaginatedResponseUtils.java` and how the `transformer` is passed.

If I had to map a collection, I'd go for stream functions such as `map`. The predicate function can be passed around, so I don't see an issue to do something similar to what it's done when calling `downloadAll`. Now there might be a good reason to not use streams that I'm not aware of.

Why implementing this `Mapper` class rather than using stream functions?

bmonjoie replied:
See answer above

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/response/PaginatedResponse.java:
```diff
@@ -0,0 +1,47 @@
+
+public class PaginatedResponse<T> {
+
+    public static boolean hasNextLink(@NonNull JSONObject jsonObject) throws JSONException {
```
FYI, the total count of entries that are paginated `.total` is also present in the response.

bmonjoie replied:
But I don't know how many have already been returned; therefore, the total is useless.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/request/BaseFHIRGetRequest.java:
```diff
@@ -0,0 +1,49 @@
+                if (bodyString == null) {
+                    throw new RuntimeException("Call was successful but body was null");
+                }
+                return fromJson(new JSONObject(bodyString));
```
suggestion:
```java
return fromJson(
  new JSONObject(
    Optional.ofNullable(body)
     .orElseThrow(() -> new Runtime("Call was successful but body was null")
     .string()
  ));
```

bmonjoie replied:
Optional requires Java8, see previous comment about that (streams)

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/request/BaseFHIRRequest.java:
```diff
@@ -0,0 +1,42 @@
+
+    @NonNull
+    protected Request.Builder getRequestBuilder(@Nullable Map<String, String> queryParameters) {
+        Request.Builder builder = new Request.Builder();
+        HttpUrl.Builder urlBuilder = Objects.requireNonNull(HttpUrl.parse(BASE_URL + endpoint + "/")).newBuilder();
+        if (queryParameters != null) {
+            for (Map.Entry<String, String> entry : queryParameters.entrySet()) {
+                urlBuilder.addQueryParameter(entry.getKey(), entry.getValue());
+            }
+        }
+        return builder.url(urlBuilder.build())
+                .addHeader("Content-Type", "application/json");
+    }
```
suggestion (not tested):

```java
HttpUrlBuilder urlBuilder = Optional
  .ofNullable(queryParameters)
  .orElse(new HashMap<String, String>())
  .entrySet()
  .stream()
  .reduce(
    Objects.requireNonNull(HttpUrl.parse(BASE_URL + endpoint + "/")).newBuilder(),
    (urlBld, entry) -> urlBld.addQueryParameter(entry.getKey(), entry.getValue()));
return (new Request.Builder()).url(urlBuilder.build()).addHeader("Content-Type", "application/json");
```

bmonjoie replied:
See comment above

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/request/BaseGraphQLRequest.java:
```diff
@@ -0,0 +1,56 @@
+            }
+        });
+
+        while (exceptions[0] == null && responses[0] == null) {
+            Thread.sleep(100);
+        }
```
If possible I'd find another way than using a `Thread.sleep` (how does it affect the Android app? is it already in a Bg service?) especially when looping on captured references (it feels like a good recipe for a deadlock). Is there a reason you don't want to go for async future execution?

bmonjoie replied:
This must be run in a `Thread`; it seems I forgot to add the  `@WorkerThread` annotation to indicate it (though the name is clear). I had to do that because Apollo doesn't expose any synchronous code. I could have used a Mutex, but it could have lead to a dead lock (if the asynchronous code finishes before the `wait` is called. I found this the cleaner solution to make the code synchronous.

I wanted the code synchronous to change as little code as possible in the original code base.

toch replied:
I like the new version :+1:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/request/PostNewClaimRequest.java:
```diff
@@ -0,0 +1,209 @@
+
+import java.util.Objects;
+
+public class PostNewClaimRequest extends BaseFHIRPostRequest<PendingClaim, String> {
```
My main concern with its implementation is that it takes of a lot of concerns as it needs to know the internals of many domain objects. As you did with your DTOs, I'd extract all the conversion in dedicated interface to isolate their procressing knowledge.

bmonjoie replied:
There is already a lot of logic in the DTO regarding how to transform it from the original JSON.
It seemed to me that putting this logic here made the most sense as I didn't forsee reusing any of this logic anywhere else.

toch replied:
we discussed that live on Friday, it's fine by me, especially that we don't want to invest that much on the present code base.

toch reviewed (commented):
@bmonjoie Here a first review mainly on `network`, so you don't have to wait I've finished to read everything. I'm reviewing the rest.

I hope to have expressed clearly, smoothly, and politely my questions, suggestions, and concerns. I consider that I don't have all the cards in hand as you do, I hope this will help.

toch commented on claimManagement/src/main/graphql/org.openimis.imisclaim/schema.graphqls:
```diff
@@ -0,0 +1,8067 @@
+type Query {
```
I'm curious.

Does it help you in the IDE? or is it used only by Apollo?
How do you keep this updated? manually?

bmonjoie replied:
It doesn't help the IDE directly; Apollo generates objects based on that, which helps the IDE.
I made the query using GraphiQL and copied/pasted it in there. The IDE has a plugin to edit it directly inside it but I found it easier with GraphiQL

toch replied:
TIL, :bow:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/ImisActivity.java:
```diff
@@ -233,6 +235,8 @@ private AlertDialog showLoginDialogBox(Runnable onLoggedIn, Runnable onCancel) {
         return alertDialogBuilder.show();
     }
 
+    protected void onUserLoggedIn(){}
```
is it required by the inherited class interface?

bmonjoie replied:
It's a way for me to have a method called when the user is logged in, which the sub-classes can override to take action. Since the base class doesn't have to do anything on login, the method here is empty.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/SQLHandler.java:
```diff
@@ -74,11 +74,14 @@ public void onCreate(SQLiteDatabase db) {
-            db.execSQL("ATTACH DATABASE '" + DB_NAME_MAPPING + "' AS dbMapping1");
-            Cursor c = db.rawQuery("select I.code,I.name,M.Type AS isMapped FROM tblReferences I LEFT OUTER JOIN dbMapping1.tblMapping M ON I.Code = M.Code WHERE I.Type =?", new String[]{Type});
-            return c;
+            try {
+                db.execSQL("ATTACH DATABASE '" + DB_NAME_MAPPING + "' AS dbMapping1");
+            } catch (SQLException ignored) {
+            }
```
what's the purpose of this statement?

is it to "connect" to the database? and you get exception if it's already the case (and that's why you nesting the try-catch)?

bmonjoie replied:
I have no idea; it was already there. I just added the try/catch because it could fail when it was already attached.

toch replied:
there might be a way to check the current attached db. anyway, no that important.

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/SQLHandler.java:
```diff
@@ -74,11 +74,14 @@ public void onCreate(SQLiteDatabase db) {
+                db.execSQL("ATTACH DATABASE '" + DB_NAME_MAPPING + "' AS dbMapping1");
+            } catch (SQLException ignored) {
+            }
+            return db.rawQuery("select I.code,I.name,M.Type AS isMapped FROM tblReferences I LEFT OUTER JOIN dbMapping1.tblMapping M ON I.Code = M.Code WHERE I.Type =?", new String[]{Type});
```
```suggestion
            return db.rawQuery("SELECT I.code, I.name, M.Type AS isMapped FROM tblReferences I LEFT OUTER JOIN dbMapping1.tblMapping M ON I.Code = M.Code WHERE I.Type =?", new String[]{Type});
```

bmonjoie replied:
I didn't want to change the original SQL, even for formatting

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/Token.java:
```diff
@@ -12,6 +12,14 @@
 public class Token {
+    private static final SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSSSXXX", Locale.US);
+
+    // It's inefficient but it's to stay backward compatible without some logic to migrate the old
```
I'd create a ticket/issue to document that possible future improvement.

bmonjoie replied:
There are many more pressing things than that one to improve. Also, this class as a whole should be rewritten :shrug:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/domain/entity/Policy.java:
```diff
@@ -0,0 +1,439 @@
+    protected Policy(Parcel in) {
+        code = Objects.requireNonNull(in.readString());
+        name = Objects.requireNonNull(in.readString());
+        if (in.readByte() == 0) {
+            value = null;
+        } else {
+            value = in.readDouble();
+        }
```
suggestion:

```java
value = in.readByte() != 0 ? in.readDouble() : null;
```

or extract in an eponym private method to make the code more readable

```java
private void readValue(Parcel in) {
  value = in.readByte() != 0 ? in.readDouble() : null; 
}
// ...
protected Policy(Parcel in) {
  // ...
  readValue();
  // ...
}
```

or add method like:

```java
private Double readNullableDouble(Parcel in) {
  return in.readByte() != 0 ? in.readDouble() : null; 
}
```

bmonjoie replied:
The code is generated by Android Studio so I didn't improve it :sweat_smile:

toch replied:
> Generators are hell

:see_no_evil: 

so :+1:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/usecase/FetchInsureeInquire.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Nullable
+    private String getPhotoPath(@NonNull List<GetInsureeInquireQuery.Photo> photos) {
+        for (GetInsureeInquireQuery.Photo photo : photos) {
+            String filename = photo.filename();
+            if (filename != null) {
+                String folder = photo.folder();
+                if (folder != null) {
+                    folder = folder.replace('\\', '/');
+                    if (!folder.endsWith("/")) {
+                        folder += "/";
+                    }
+                    return folder + filename;
+                }
+                return filename;
+            }
+        }
+
+        return null;
```
suggestion

```java
return photos
  .stream()
  .filter(photo -> photo.filename() != null)
  .map(photo -> photo.filename())
  .findFirst()
  .ifPresent(filename -> {
    return Optional.ofNullable(filename.folder())
      .ifPresent(folder -> {
        folder = folder.replace('\\', '/');
        folder += !folder.endsWith("/") ? "/" : "";
        return folder;
      })
      .orElse("")
      + filename;
  })
  .orElse(null);
```

bmonjoie replied:
Same as above

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/usecase/FetchInsureeInquire.java:
```diff
@@ -0,0 +1,125 @@
+     * <a href="https://github.com/openimis/database_ms_sqlserver/blob/main/sql/stored_procedures/uspAPIGetCoverage.sql#L48C3-L48C101">Values for the status in the stored procedure</a>
+     */
+    @NonNull
+    private Policy.Status intAsStatus(@Nullable Integer integer) {
```
I'd put that in the enum class `Policy.Status`.

bmonjoie replied:
I don't want my domain object to know how it is mapped into/from the API response. Therefore, that's why it's here and not in the domain object.

toch replied:
we discussed that live on Friday. So :+1:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/usecase/FetchClaims.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Nullable
+    private Integer statusAsInt(@Nullable Claim.Status status) {
```
I'd put that in the enum directly

bmonjoie replied:
Same as above

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/usecase/FetchClaims.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Nullable
+    private Claim.Status intAsStatus(@Nullable Integer status) {
```
I'd put that in the enum `Status`.

bmonjoie replied:
Same as above

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/domain/entity/PendingClaim.java:
```diff
@@ -0,0 +1,423 @@
+        public static List<Service> fromJson(@NonNull JSONArray array) throws Exception {
+            List<Service> list = new ArrayList<>();
+            for (int i = 0; i < array.length(); i++) {
+                JSONObject medication = array.getJSONObject(i);
+                list.add(new Service(
+                        /* code = */ medication.getString("ServiceCode"),
+                        /* price = */ Double.parseDouble(medication.getString("ServicePrice")),
+                        /* quantity = */ Integer.parseInt(medication.getString("ServiceQuantity"))
```
as we're dealing with `Service`:

```suggestion
                JSONObject service = array.getJSONObject(i);
                list.add(new Service(
                        /* code = */ service.getString("ServiceCode"),
                        /* price = */ Double.parseDouble(service.getString("ServicePrice")),
                        /* quantity = */ Integer.parseInt(service.getString("ServiceQuantity"))
```

bmonjoie replied:
Copypasta gone wrong :sweat_smile:

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/domain/entity/PendingClaim.java:
```diff
@@ -0,0 +1,423 @@
+        };
+    }
+
+    public static class Medication implements Parcelable {
```
why not using the existing class? they look alike.

bmonjoie replied:
It has different properties

toch reviewed (commented):
second review, focused on domain and usecase. there is still `java.org.openimis.imisclaims` to read.

bmonjoie committed:
PR comments

toch commented on claimManagement/src/main/java/org/openimis/imisclaims/network/request/BaseGraphQLRequest.java:
```diff
@@ -29,23 +35,26 @@ public abstract class BaseGraphQLRequest {
-        while (exceptions[0] == null && responses[0] == null) {
-            Thread.sleep(100);
+        if (!semaphore.tryAcquire(TIME_OUT_IN_MS, TimeUnit.MILLISECONDS)) {
+            throw new TimeoutException("Call couldn't finish within " + TIME_OUT_IN_MS + "ms");
```
:+1:

toch commented:
I've reviewed your answers. I can't resolve the discussions :shrug: so I've :+1: or answered them to do so.

bmonjoie committed:
Fix issues after first review

bmonjoie committed:
Detect "REJECTED" status

Based on the method `build_fhir_total_list` in the Python code,
if there is no total, the claim is rejected.

bmonjoie committed:
Fix error message when Insuree is not found

dragos-dobre commented:
@bmonjoie We don't integrate code directly to main branch. Our development workflow follows GitFlow. Please create a PR for develop branch. I'm closing this PR.

bmonjoie commented:
@dragos-dobre I have reopened the PR and changed the destination branch not to lose the history
